File: widths.lua
local widths = {}

local math = require "math"
local max = math.max
local pairs = pairs

local function glyphw(glyph)
  local width = 0
  for _,path in pairs(glyph) do
    for _,point in pairs(path) do
      width = max(width,point[1])
    end
  end
  return width
end

widths.glyph = glyphw

function widths.glyphs(glyphs)
  local widths_t = {}
  for k, glyph in pairs(glyphs) do
    widths_t[k] = glyphw(glyph)
  end
  return widths_t
end

return widths

File: widths.js
function setWidthFromPaths(glyph){
  var width = 0;
  for (var i = 0; i < glyph.paths.length; i++){
    for (var j = 0; j < glyph.paths[i].length; j++){
      width = Math.max(width,glyph.paths[i][j][0]);
    }
  }
  glyph.width = width;
  return glyph;
}

function setWidthsForFont(font){
  if(font.glyphs) {
    for (var glyph in font.glyphs) {
      if(!font.glyphs[glyph].width && font.glyphs[glyph].paths)
      setWidthFromPaths(font.glyphs[glyph]);
    }
  }
  if(font.noglyph && !font.noglyph.width && font.noglyph.paths) {
    setWidthFromPaths(font.noglyph);
  }
  return font;
}

exports.setWidthFromPaths = setWidthFromPaths;
exports.setWidthsForFont = setWidthsForFont;